internal/core/services/events_forwarder/events_forwarder_service.go

```diff
@@ -24,60 +24,171 @@ package events_forwarder
+	instanceStorage  ports.GameRoomInstanceStorage
+	schedulerCache   ports.SchedulerCache
 }
 
-func NewEventsForwarderService(eventsForwarder ports.EventsForwarder) interfaces.EventsService {
+func NewEventsForwarderService(
+	eventsForwarder ports.EventsForwarder,
+	schedulerStorage ports.SchedulerStorage,
+	instanceStorage ports.GameRoomInstanceStorage,
+	schedulerCache ports.SchedulerCache,
+) interfaces.EventsService {
 	return &EventsForwarderService{
 		eventsForwarder,
 		zap.L().With(zap.String("service", "rooms_api")),
+		schedulerStorage,
+		instanceStorage,
+		schedulerCache,
 	}
 }
 
 func (es *EventsForwarderService) ProduceEvent(ctx context.Context, event *events.Event) error {
-	room := &game_room.GameRoom{}
-	instance := &game_room.Instance{}
-
-	switch event.Name {
-	case events.RoomEvent:
-		return es.forwardRoomEvent(ctx, room, instance, map[string]interface{}{}, "")
-	case events.PlayerEvent:
-		return es.forwardPlayerEvent(ctx, room, map[string]interface{}{}, "")
+	if _, ok := event.Attributes["eventType"]; !ok {
+		return errors.New("eventAttributes must contain key \"eventType\"")
 	}
-	return nil
+	eventType := event.Attributes["eventType"].(string)
 
+	scheduler, err := es.getScheduler(ctx, event.SchedulerID)
+	if err != nil {
+		return err
+	}
+
+	forwarderList := scheduler.Forwarders
+	if len(forwarderList) > 0 {
+		for _, _forwarder := range forwarderList {
+			switch event.Name {
+			case events.RoomEvent:
+				instance, err := es.instanceStorage.GetInstance(ctx, event.SchedulerID, event.RoomID)
+				if err != nil {
+					es.logger.Error(fmt.Sprintf("Failed to get instance for room \"%v\" from scheduler \"%v\" info", event.RoomID, event.SchedulerID), zap.Error(err))
+					return err
+				}
+				err = es.forwardRoomEvent(ctx, event, eventType, *instance, *scheduler, _forwarder)
+				if err != nil {
+					return err
+				}
+			case events.PlayerEvent:
+				err = es.forwardPlayerEvent(ctx, event, eventType, _forwarder)
+				if err != nil {
+					return err
+				}
+			}
+		}
+	} else {
+		es.logger.Info(fmt.Sprintf("scheduler \"%v\" do not have forwarders configured", event.SchedulerID))
+	}
+
+	return nil
 }
 
-func (es *EventsForwarderService) forwardRoomEvent(ctx context.Context, room *game_room.GameRoom, instance *game_room.Instance, attributes map[string]interface{}, options interface{}) error {
-	err := es.eventsForwarder.ForwardRoomEventObsolete(ctx, room, instance, attributes, options)
+func (es *EventsForwarderService) forwardRoomEvent(
+	ctx context.Context,
+	event *events.Event,
+	eventType string,
+	instance game_room.Instance,
+	scheduler entities.Scheduler,
+	_forwarder *forwarder.Forwarder,
+) error {
+	if len(instance.Address.Ports) == 0 {
+		return fmt.Errorf("no room port found to forward roomEvent. Forwarder name: \"%v\", Scheduler: \"%v\"", _forwarder.Name, event.SchedulerID)
+	}
+	selectedPort := instance.Address.Ports[0].Port
+
+	roomEvent := events.RoomEventType(eventType)
```

Comment:
@caiordjesus you can pass any string for this eventType value and the code will not break, what about creating and using some methods for converting enums instead of using weak typecast here? We can use something like the convert methods from GameRoomPingStatus internal/core/entities/game_room/game_room.go:103, what do you think?

Reply:
Nice, I'll make this improve right now!